accepted/2021/runtime-security-mitigations.md

```diff
@@ -0,0 +1,213 @@
+
+The security mitigations prevent or reduce probability of each step in this sequence.
+
+The mitigations for memory safety violations must be typically applied to all code running in a process to be fully effective. The exploits often combine memory safety violations and code from multiple modules in the process. For example, memory safety violation in C/C++ module and code generated by .NET runtime JIT can be combined together to produce an exploit. Thus, missing mitigations in .NET modules can allow exploiting vulnerabilities in C/C++ modules that would be hard or impossible to exploit in isolation, and vice versa.
+
+# Executable Space Protection
+
+_Known as Data Execution Prevention (DEP) on Windows_
+
+The simplest path for an attack is to write new malicious code in memory and then get control flow to jump to it. Executable space protection enforces that memory has independent write and execution permissions. Attacks can no longer create new code because executable memory is not writable and writable memory is not executable.
+
+## W^X
+
+[W^X](https://en.wikipedia.org/wiki/W%5EX) is one of the most fundamental mitigations. It blocks the simplest attack path by disallowing memory pages to be writeable and executable at the same time. .NET runtime has been missing this mitigation so far and the lack of this mitigation has (correctly) resulted in us not considering more advanced mitigations. Large number of pages that are both writeable and executable in a typical .NET process is a ripe target for attacks that simply inject new code.
+
+Apple has made the W^X mandatory for future versions of macOS desktop operating system as part of Apple Silicon transition. It motivated us to schedule implementation of this mitigation for .NET 6, on all supported operating systems. Our principle is to treat all supported operating systems equally with respect to security, where possible.
+
+## Protect Intermediate Language (IL) Code
+
+IL code is equivalent of executable code. The IL code might be overwritten with malicious IL. An attacker could then wait for the IL to be compiled or re-compiled via tiering and have malicious code injected that way.
+
+.NET runtime should treat the IL code same way as directly executable code, ie make IL read-only or otherwise protected wherever possible.
+
+## Address Space Layout Randomization (ASLR)
+
+[Address space layout randomization](https://en.wikipedia.org/wiki/Address_space_layout_randomization) arranges all code and data structures in memory randomly that makes it harder for the attacker to reliably transfer control to particular exploited function or modify particular data structure. .NET runtime has been compatible with ASLR since .NET Framework 3.5 SP1 (released in 2008).
```

Comment:
I assume if the customer app loads native libraries that aren't ASLR aware, the OS loader still supports ASLR for our runtime.

Do we verify that all our native code (beyond coreclr/jit/gc) opt into ASLR? I guess this falls into the category of outer ring checks we do.

Reply:
> the customer app loads native libraries that aren't ASLR aware, the OS loader still supports ASLR for our runtime.

Yes, it does. Note that the non-ASLR enabled library can be used as trampoline to attack anything else in the process, including our runtime. It is really important for these mitigations to be enabled for all code in the process as discussed in the "Anatomy of an exploit" chapter.

> Do we verify that all our native code (beyond coreclr/jit/gc) opt into ASLR?

Yes, it is part of the binskim gate. It actually found some Linux binaries that were not opted in correctly when it was enabled for the first time.

Reply:
Naïve question, is there anything we can do to help customers realize when they load native code into a .NET process that isn't ASLR enabled? Or if not technically at least educate through documentation.